src/main/java/com/pusher/client/channel/PusherEvent.java

```diff
@@ -1,45 +1,78 @@
 package com.pusher.client.channel;
 
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.annotations.SerializedName;
+
 import java.util.Map;
 
 public class PusherEvent {
-    private Map<String, Object> eventData;
-
-    public PusherEvent(Map<String, Object> eventData) {
-        this.eventData = eventData;
-    }
-
-    /**
-     * getProperty returns the value associated with the key, or null.
-     * It is recommended that you use the specialized getters in this class instead.
-     *
-     * @param key The key you wish to get.
-     * @return
-     *      The object can be casted as follows:
-     *      - JSON strings - java.lang.String
-     *      - JSON number - java.lang.Double
-     *      - JSON boolean - java.lang.Boolean
-     *      - JSON array - java.util.ArrayList
-     *      - JSON object - java.util.Map
-     *      - JSON null - null
-     */
-    public Object getProperty(String key) {
-        return eventData.get(key);
-    }
-
-    /**
-     * Returns the userId associated with this event.
-     *
-     * @return
-     *      The userID string: https://pusher.com/docs/channels/using_channels/events#user-id-in-client-events,
-     *      or null if the event is not a client event on a presence channel.
-     */
-    public String getUserId() { return (String)eventData.get("user_id"); }
-    public String getChannelName() { return (String)eventData.get("channel"); }
-    public String getEventName() { return (String)eventData.get("event"); }
-    public String getData() { return (String)eventData.get("data"); }
+
+    @SerializedName("user_id")
+    private final String userId;
+
+    private final String data;
+    private final String channel;
+    private final String event;
+
+    public String getUserId() {
+        return userId;
+    }
+
+    public String getChannelName() {
+        return channel;
+    }
+
+    public String getEventName() {
+        return event;
+    }
+
+    public String getData() {
+        return data;
+    }
```

Comment:
Are the changes in this class needed for the fix? This is a breaking change in the public interface.

Reply:
I've added the getProperty method back, so now the public API is the same?

Reply:
Yes. That's better. I would also remove the automatic conversion from object to string on the data field.

All our SDKs should be serializing event data as strings, but that's not a breaking change of course.

Reply:
> I would also remove the automatic conversion from object to string on the data field.
I'm not sure what you mean here, it was a string and it's still a string?

Reply:
I'm referring to the `fromJson` method.

Reply:
Aha, yes. The conversion is because some unit test provide valid json as data, while others provide a json string, for example:
in testDataIsExtractedFromMessageAndPassedToSingleListener:
`PusherEvent.fromJson("{\"event\":\"" + EVENT_NAME + "\",\"data\":\"{\\\"fish\\\":\\\"chips\\\"}\"}")`
but in testReceiveMessageAfterUnsubscribeDoesNotPassItToChannel:
`PusherEvent.fromJson("{\"event\":\"my-event\",\"data\":{\"fish\":\"chips\"},\"channel\":\"" + CHANNEL_NAME + "\"}")`
Where Gson is encoding is as a Map

Reply:
Ah, I see. Test data is incorrect then. Let's fix that and make `PusherEvent` simpler

Reply:
Well the data can be anything (even not json).

Reply:
It should be always be a string.

Reply:
that is true, I'll fix the unit tests and always handle it as string

Reply:
done